Realm/Realm/Dynamic/MetaRealmObject.cs

```diff
@@ -345,11 +279,26 @@ public override DynamicMetaObject BindSetMember(SetMemberBinder binder, DynamicM
 
             arguments.Add(valueExpression);
 
-            var expression = Expression.Block(Expression.Call(Expression.Field(GetLimitedSelf(), RealmObjectObjectHandleField), setter, arguments), Expression.Default(binder.ReturnType));
+            var expression = Expression.Block(Expression.Call(Expression.Field(self, RealmObjectObjectHandleField), setter, arguments), Expression.Default(binder.ReturnType));
+            return new DynamicMetaObject(expression, GetBindingRestrictions(self));
+        }
 
+        private BindingRestrictions GetBindingRestrictions(Expression self)
+        {
             var argumentShouldBeDynamicRealmObject = BindingRestrictions.GetTypeRestriction(Expression, _metadata.Schema.IsEmbedded ? typeof(DynamicEmbeddedObject) : typeof(DynamicRealmObject));
-            var argumentShouldBeInTheSameRealm = BindingRestrictions.GetInstanceRestriction(Expression.Field(GetLimitedSelf(), RealmObjectRealmField), _realm);
-            return new DynamicMetaObject(expression, argumentShouldBeDynamicRealmObject.Merge(argumentShouldBeInTheSameRealm));
+            var argumentShouldBeInTheSameRealm = BindingRestrictions.GetInstanceRestriction(Expression.Field(self, RealmObjectRealmField), _realm);
+            var argumentShouldBeTheSameType = BindingRestrictions.GetExpressionRestriction(
+                Expression.Equal(
+                    Expression.Constant(_metadata.Schema.Name),
+                    Expression.Property(
+                        Expression.Field(
+                            Expression.Field(
+                                self,
+                                RealmObjectMetadataField),
+                            ObjectMetadataSchemaField),
+                        SchemaGetNameProperty)));
```

Comment:
This is the real fix for the dynamic tests - it creates an expression restriction that will prevent the runtime from caching the set/get member if the schema name is different (i.e. if the dynamic object represents a different object type).